### ISSUE TITLE
Add a local NONG list and a NONG finder

### ISSUE THREAD

**Issue by xparadoxical** [Fleeym/nongd#8]
I have a lot of downloaded songs, so checking each one for whether it's a NONG to upload them to the Song File Hub afterwards would be very time-consuming. Why not automate this? People might have many NONGs sitting there, so encourage them to protect these files from being lost.

First, there needs to be a local NONG list. It'd be accessible from the Downloaded Songs list with a button. The new list would preferably look like the Downloaded Songs list menu. Initially, the list would have to be empty. Next to the NONG list, there'd be the scan button.

The scan button (+ a confirmation) would initiate a scan. With respect to the GD server rate limits, every local song would be checked by:
1. sending a request to `getGJSongInfo.php`,
2. checking if the response is `-2` (not allowed for use, NONG) or `-1` (doesn't exist, NONG);
3. if it's neither of the above, check if field 10 (download URL) contains the domain:
   - `geometrydashcontent.b-cdn.net` (archived, NONG),
   - `audio.ngfiles.com` (available, not NONG),
   - neither of the above (probably it's a 2.2 SFX);

The list should also let me copy the received link to use it in a SFH submission, if available.

**Comment by Fleeym** [comment]
Hey, thanks a lot for your suggestion! I'm guessing you're talking about songs that you already manually swapped out.

A scanning feature would get you rate limited instantly probably (there's a lot of songs). I had an idea in mind to detect if your song has been replaced based on filesize (I know, not the greatest idea), or maybe metadata. Also if the request fails then yeah its definetly a NONG. You could then swap it out with the default song, and the replaced song becomes a selectable NONG.

Not sure what you mean by copying the received link. It's not needed to upload stuff from Rob's CDN to SFH.

**Comment by xparadoxical** [comment]
No, I'm talking about songs that were taken down from NG but still exist on a player's drive. Imagine downloading a level with its song but after some time the song got taken down. It could be uploaded to SFH.

GD rate limits can be and have been found by experimentation. There's info on the GD Programming discord, [the rate limit should be 2 rps](https://discord.com/channels/646101505417674758/651480005536383009/1170012644971905155). IIRC first you get a few completely invalid responses from GD servers and if you continue, you get tempblocked by Cloudflare with http 429 and Retry-After. Checking e.g. 1000 songs (that's how many I have) would take at least 500s or 8min20s, but there's an additional layer of rate limiting so it could be longer.

A song link is required in #nong-song-submissions on the SFH discord, which is why I proposed link copying. I don't know what's possible with the SFH website and API (if there's one).

**Comment by Fleeym** [comment]
I see. I'm still not agreeing with scanning your whole song library, it would be fine to scan when you interact with the song (example entering a level page that uses it, selecting it in the editor, etc).

If the song is taken down while you have it on your device, the URL to download it still becomes unavailable, so you can't really use that in a request.

**Comment by xparadoxical** [comment]
Yeah, link copying makes sense only when it's a URL to the GD CDN.

I don't know *why* you're disagreeing. From the info I found, currently [there isn't a way to be automatically banned](https://discord.com/channels/646101505417674758/651480005536383009/1112238350762520576) for exceeding rate limits, being rate limited can be detected, you can also send at a lower rate just to be careful. Checking a single song in a level page and in level settings in the editor would be something already, but for players (especially those who don't know how to automate this themselves) who have quite a higher number of songs than 1, finding HQ songs they have that are not available online would be too time consuming to consider doing. The longer these files sit on drives and collect dust, the higher the chance of something happening to them becomes - like a game reinstall/repair or a system reinstall. I imagine there are less popular songs used in rated levels but not archived on SFH that could become extinct due to: 1. no video containing the song existing on internet, and people who care about NONGs being unable to find anyone that'd 2. care to check if they have the song and 3. have the song. That's why I think it's best to do what you can to preserve them.

**Comment by Fleeym** [comment]
I see. The main problem then is somehow making the process of submitting it to SFH easy. I can talk with the guys that run the API, see if they like the idea.

Isn't copying links from the GD CDN pointless? Those don't need to be posted to SFH as far as I know.

**Comment by Fleeym** [comment]
Good news, I had a talk with the API guy, we figured most of it out theoretically.
After 2.2 comes out we'll take a stab at it :)
